### DISCUSSION TITLE
i64x4 and avx 512

### DISCUSSION THREAD

**Discussion by mindbeast** [ispc/ispc#2170] [Q&A] [answered]
I have a double precision workload that seems to significantly benefit from the lower register pressure of the i64x4 avx2 target. An equivalent target that allows the same 256 bit vector width, but that allows access to the full set of 32 vector registers (as opposed to just the 16 that avx2 allows) and avx512 mask registers, seems like it would be very useful. I can't see it in the existing targets, though. Is there any way for me to get what I want, or should I turn this into a feature request?

I can try the avx512skx-i32x8 target and use a full 512 bit vector, but I expect the [frequency penalty](https://travisdowns.github.io/blog/2020/08/19/icl-avx512-freq.html) that will be paid on SKX will prevent me from using it.

**Comment by dbabokin** [comment, marked as answer]
This is very reasonable request and I've already heard it from other developers.

One caveat here, do not take `i32` or `i64` in AVX512 targets seriously. We originally had this "base type" in the target naming schema because it affects how the default mask would look like on the targets that don't have mask registers. I.e. for SSE2/4 and AVX2 it matters a lot. But for AVX512 we have mask registers and `avx512skx-i32x4` and `avx512skx-i64x4` would be the same thing. Just keep that in mind.

As for the how to proceed with adding `avx512skx-i32x4` target, let's convert this question to feature request and I'll implement it.

**Reply by mindbeast** [reply]
It wasn't clear to me at all why prior targets had a base type + the program count until your explanation just now. Might be good for that to be in documentation somewhere, it's fairly easy for users to not be aware that there is some configurability in program count when first using ISPC.

I also want to say thanks for being really responsive and helpful. I'll file that feature request ASAP.

**Reply by mindbeast** [reply]
https://github.com/ispc/ispc/issues/2171

**Comment by dbabokin** [comment]
> Might be good for that to be in documentation somewhere

Yep. I'm planning to do that along with allowing to drop specifying base type in the targets (at least for the targets that don't care about it, i.e. AVX512 and GPU targets).
